src/qibo/abstractions/states.py

```diff
@@ -0,0 +1,267 @@
+        return self._nqubits
+
+    def __len__(self):
+        """Number of components in the state's tensor representation."""
+        if self._nqubits is None:
+            raise_error(AttributeError, "State number of qubits not available.")
+        return self.nstates
+
+    @nqubits.setter
+    def nqubits(self, n):
+        self._nqubits = n
+        self.nstates = 2 ** n
+
+    @property
+    @abstractmethod
+    def shape(self): # pragma: no cover
+        """Shape of the state's tensor representation."""
+        raise_error(NotImplementedError)
+
+    @property
+    @abstractmethod
+    def dtype(self): # pragma: no cover
+        """Type of state's tensor representation."""
+        raise_error(NotImplementedError)
+
+    @property
+    def tensor(self):
+        """Tensor representation of the state in the computational basis."""
+        if self._tensor is None:
+            raise_error(AttributeError, "State tensor not available.")
+        return self._tensor
+
+    @tensor.setter
+    def tensor(self, x):
+        nqubits = int(math.log2(len(x)))
+        if self._nqubits is None:
+            self.nqubits = nqubits
+        elif self._nqubits != nqubits:
+            raise_error(ValueError, "Cannot assign tensor of length {} to "
+                                    "state with {} qubits."
+                                    "".format(len(x), self.nqubits))
+        self._tensor = x
+
+    @abstractmethod
+    def __array__(self): # pragma: no cover
+        """State's tensor representation as an array."""
+        raise_error(NotImplementedError)
+
+    @abstractmethod
+    def numpy(self): # pragma: no cover
+        """State's tensor representation as a numpy array."""
+        raise_error(NotImplementedError)
+
+    def state(self, numpy=False):
+        """State's tensor representation as an backend tensor.
+
+        Args:
+            numpy (bool): If ``True`` the returned tensor will be a numpy array,
+                otherwise it will follow the backend tensor type.
+                Default is ``False``.
+        """
+        if numpy:
+            return self.numpy()
+        return self.tensor
+
+    def __repr__(self):
+        return str(self.tensor)
+
+    def __getitem__(self, i):
+        if isinstance(i, int) and i >= self.nstates:
+            raise_error(IndexError, "State index {} out of range.".format(i))
+        return self.tensor[i]
+
+    @classmethod
+    def from_tensor(cls, x, nqubits=None):
+        """Constructs state from a tensor.
+
+        Args:
+            x: Tensor representation of the state in the computational basis.
+            nqubits (int): Optional number of qubits in the state.
+                If ``None`` it is calculated automatically from the tensor
+                representation shape.
+        """
+        obj = cls(nqubits)
+        obj.tensor = x
+        return obj
+
+    @classmethod
+    @abstractmethod
+    def zstate(cls, nqubits): # pragma: no cover
```

Comment:
Please replace `zstate` with `zero_state` or `vacuum_state` so it is easier to interpret.